### ISSUE TITLE
Assertion 'fd' failed at ../systemd/src/basic/socket-util.c:1011, function getpeergroups(). Aborting.

### ISSUE THREAD

**Issue by bl33pbl0p** [systemd/systemd#8075]
### Submission type

 - Bug report

### systemd version the issue has been seen with

237, git master

### Used distribution

Arch Linux, Exherbo Linux 

### In case of bug report: Expected behaviour you didn't see

`systemd-analyze -H user@localhost blame` works normally

### In case of bug report: Unexpected behaviour you saw

```
[root@localhost ~]# systemd-analyze -H kowalski@localhost blame
kowalski@localhost's password: 
Failed to parse reply: Connection reset by peer
```

In the logs, 
```
Feb 02 17:00:35 epsilon systemd-stdio-bridge[53]: Assertion 'fd' failed at ../systemd/src/basic/socket-util.c:1011, function getpeergroups(). Aborting.
```
This commit looks related, https://github.com/systemd/systemd/commit/43f2c88df036ba8ef435088f31617a2646062799

**Comment by bl33pbl0p** [comment]
Better logs,
```
Feb 02 11:29:37 delta systemd-stdio-bridge[16339]: Assertion 'fd' failed at ../systemd-237/src/basic/socket-util.c:1011, function getpeergroups(). Aborting.
Feb 02 11:29:37 delta audit[16339]: ANOM_ABEND auid=1000 uid=1000 gid=1000 ses=57 pid=16339 comm="systemd-stdio-b" exe="/usr/x86_64-pc-linux-gnu/bin/systemd-stdio-bridge" sig=6 res=1
Feb 02 11:29:37 delta kernel: audit: type=1701 audit(1517587177.825:13): auid=1000 uid=1000 gid=1000 ses=57 pid=16339 comm="systemd-stdio-b" exe="/usr/x86_64-pc-linux-gnu/bin/systemd-stdio-bridge" sig=6 res=1
Feb 02 11:29:37 delta systemd[1]: Started Process Core Dump (PID 16340/UID 0).
Feb 02 11:29:38 delta sshd[16338]: Received disconnect from ::1 port 46658:11: disconnected by user
Feb 02 11:29:38 delta sshd[16338]: Disconnected from user kowalski ::1 port 46658
Feb 02 11:29:38 delta sshd[16334]: pam_unix(sshd:session): session closed for user kowalski
Feb 02 11:29:38 delta systemd[1]: sshd@9-::1:22-::1:46658.service: Consumed 22ms CPU time, received 0B IP traffic, sent 0B IP traffic
Feb 02 11:29:38 delta systemd[1]: session-57.scope: Consumed 9ms CPU time, received 0B IP traffic, sent 0B IP traffic
Feb 02 11:29:38 delta systemd-logind[623]: Removed session 57.
Feb 02 11:29:38 delta systemd-coredump[16341]: Process 16339 (systemd-stdio-b) of user 1000 dumped core.
                                               
                                               Stack trace of thread 16339:
                                               #0  0x00007fb4975ad9e0 raise (libc.so.6)
                                               #1  0x00007fb4975af0ae abort (libc.so.6)
                                               #2  0x00007fb4979dcd58 log_assert_failed_realm (libsystemd-shared-237.so)
                                               #3  0x00007fb4979efc1a getpeergroups (libsystemd-shared-237.so)
                                               #4  0x00007fb497a29cfa bus_get_peercred (libsystemd-shared-237.so)
                                               #5  0x00007fb497a3460a bus_start_fd (libsystemd-shared-237.so)
                                               #6  0x00005570aac21458 main (systemd-stdio-bridge)
                                               #7  0x00007fb49759a021 __libc_start_main (libc.so.6)
                                               #8  0x00005570aac21a9a _start (systemd-stdio-bridge)
Feb 02 11:29:38 delta systemd[1]: systemd-coredump@16-16340-0.service: Consumed 55ms CPU time, received 0B IP traffic, sent 0B IP traffic
```

**Comment by anomiex** [comment]
I ran into this same assertion failure in a few different contexts, including execution of /sbin/shutdown from a daemon listening for a custom key press.

I find it suspicious that getpeersec() and getpeercred() both assert `fd >= 0`, while the new getpeergroups() is asserting `fd` instead.

**Comment by vcaputo** [comment]
I threw up this https://github.com/systemd/systemd/pull/8080 because that assert looks obviously wrong.

**Comment by simon04** [comment]
I ran into that when using `lftp` to open an FTP connection:
```
$ lftp ftp.example.com 
Assertion 'fd' failed at ../systemd-stable/src/basic/socket-util.c:1011, function getpeergroups(). Aborting.
```

(I'm posting this since it might help other people searching for a solution of this problem.)

**Comment by stuart-little** [comment]
Seconding @simon04: I'm seeing this on Arch with their latest version of systemd (237) when trying to connect to a Box account with lftp.

**Comment by piluex** [comment]
This is happening to me while using rsync locally.

> 162,623,246   1%  204.89kB/s    0:12:55 (xfr#1549, to-chk=669/269773)Assertion 'fd' failed at >../systemd-stable/src/basic/socket-util.c:1011, function getpeergroups(). Aborting.

>rsync: connection unexpectedly closed (163968 bytes received so far) [sender]
>rsync error: error in rsync protocol data stream (code 12) at io.c(226) [sender=3.1.3]

**Comment by stuart-little** [comment]
This is just a quick confirmation that [commit 75f4077](https://github.com/systemd/systemd/commit/75f40779607ea79f20441c7fb46744d04ee2c7ae) does seem to fix this: I built the three *systemd* Arch packages (systemd, libsystemd, systemd-sysvcompat) against that commit and can now use lftp as before without issue.

**Comment by presto8** [comment]
I was affected by this bug. Just curious, why are assert()s being compiled in release (non-debug) code on Arch builds of systemd?

**Comment by vcaputo** [comment]
@presto8 That seems like a question for the Arch maintainers.

**Comment by presto8** [comment]
@vcaputo Thanks, I will check with them! I can tell them it would be the recommendation of systemd developers to disable asserts (-DNDEBUG etc.) for production builds?

**Comment by vcaputo** [comment]
@presto8 Considering we benefit from the feedback reports from those asserts provide, and it isn't particularly performance sensitive code, i cannot take such a position.

It's up to the distributor to decide how to build their software for production use.

**Comment by poettering** [comment]
> I was affected by this bug. Just curious, why are assert()s being compiled in release (non-debug) code on Arch builds of systemd?

I'd advise against turning off asserts... This kind of optimization is unlikely to really get you much performance benefit as asserts are seldom used in inner loops. Moreover pretty often it is better idea to trigger an actionable assertion so that things can get fixed rather stay hidden silently.

Unless you are in extremely resource constrained systems I am pretty sure turning off assertions will do more harm than good.

**Comment by piluex** [comment]
Hello, 

As an user this kind of assertions are blocking my backup scripts which just use rsync locally... I don't think that's a complex or strange scenario.

I don't think that arch users (including me) should get rsync and lftp broken thanks to systemd assertions, neither regular users will be reading those and sending proper bug reports. Specially when the assertion is just failing in a nonsense way. 

To be honest I didn't spend much time into the issue to give a more technical point of view. But from the user perspective this particular assertion doesn't seem to make much sense. I updated systemd and now rsync is broken for a very VERY simple use case.

Is there a 'hidden' bug to be fixed? Maybe, but rsync was working fine before, and it's not working now.

I will try to make some time to build this myself and print some useful information with the assertion, so at least there's something to win from this. Hopefully I don't break anything important while trying on my backup-less environment. 

I'm happy to help but please consider that the affected people reporting this issues do need their systems working to report them, and I don't remember singing up for a systemd beta testing program when installing Arch, so please don't make decissions in the name of Arch users.

**Comment by vcaputo** [comment]
@piluex FYI https://wiki.archlinux.org/index.php/downgrading_packages

**Comment by piluex** [comment]
@vcaputo Thanks, I've on a side-note already downloaded the repo and trying to build the latest version to test. I will now move to  the irc channel to avoid future noise in here.

**Comment by lilydjwg** [comment]
I see this was fixed in #8080, but is there a test covering this? I didn't see any related updates with the existing tests.
